### ISSUE TITLE
Watch mode interacts poorly with Angular change detection

### ISSUE THREAD

**Issue by wclaibor** [jhildenbiddle/css-vars-ponyfill#124]
I'm using css-vars-ponyfill in my Angular application and ran into an issue when using watch mode. It seems that Angular change detection is creating an infinite loop in the MutationObserver. I believe this could be fixed by allowing users to attach the MutationObserver to the head element rather than the document element.

See the attached stackblitz for an example Angular application that triggers this infinite loop.
https://stackblitz.com/edit/css-vars-ponyfill-performance?file=src%2Fapp%2FponyfillCssVars.ts

**Comment by Phil147** [comment]
Hi @wclaibor 
I don't know exactly why it is happening in APP_INITIALIZER that it crashes but using the ponyfill in the `polyfills.ts` seems fine. We are also using the ponyfill conditionally in an angular component during angulars change detection to turn it off and on and didn't have any issues so far. (At least with version 2.2.1 have to check the latest)

**Comment by wclaibor** [comment]
@Phil147 interesting wrinkle. I'm working on an Angular library, so I really do need to use an APP_INITIALIZER. Not sure why that would cause a problem when everything works in the polyfill though.

**Comment by jhildenbiddle** [comment]
@wclaibor --

The ponyfill's observer acts only on specific `<link>` and `<style>` mutations, so my first guess is that your app (or Angular itself) is making `<link>` and/or `<style>` mutations after each ponyfill update. It's certainly possible this isn't the case and there's a bug in the ponyfill, but I think the first step to debugging is to determine what mutations are occurring and which library/framework is initiating them.

I took a look at your demo and I'm not seeing the issue--probably because I'm not using your demo properly. Here's what I did:

1. Added `onlyLegacy: false` to the `cssVars()` call in `ponyfillCssVars.ts`. This is just so I can test in a modern browser. The ponyfill behaves the same in modern and legacy browsers when `onlyLegacy: false`, but perhaps this convenience is throwing off the test.

1. Removed the `setTimeout()` wrapper around `cssvars()` in `ponyfillCssVars.ts`.

1. Uncommented lines 18-22 in `app.module.ts`

If you can provide instructions for how to get your demo to present the issue described, I can be of more assistance (although I'm not an Angular guy, so my ability to help may be limited).

Perhaps setting up another MutationObserver to console.log mutations data would be a good place to start. The observer would be chatty initially, but you can filter as needed until it can be determined exactly which mutations are causing the loop. Alternatively, you can edit the ponyfill source to see what mutations it's acting on (see here: https://github.com/jhildenbiddle/css-vars-ponyfill/blob/b62260fb157e03402e602532273369d4d349f7e0/src/index.js#L709-L726)

Hope this helps!

**Comment by jhildenbiddle** [comment]
Just thought of one more possibility...

In 2.3.0, a new behavior was introduced that disabled source `<link>` and `<style>` nodes after the ponyfill has transformed CSS and appended it to the DOM.

For example, assume you have this `<style>` element:

```html
<style>
:root { 
  --color: red; 
}
p { 
  color: var(--color); 
}
</style>
```

After the ponyfill runs, you'll have the following:

```html
<style data-cssvars="src">
:root { 
  --color: red; 
}
p { 
  color: var(--color); 
}
</style>

<style data-cssvars="out">
p { 
  color: red; 
}
</style>
```

In versions up to 2.2.1, the browser had to parse both stylesheets even though the first one is being overridden by the second. In 2.3.0, the first stylesheet is disabled after the ponyfill appends new CSS. This reduces the amount of CSS the browser has to parse by 50% which can have a dramatic impact on performance.

Perhaps your app (or Angular) is detecting source stylesheets being disabled by the ponyfill, making mutations it believes are necessary as a result, which creates the loop with the ponyfill observer.

TL;DR: Try 2.2.1 and see what happens. :)

**Comment by wclaibor** [comment]
@jhildenbiddle, thanks for the quick response on this.

The problem I am seeing does not involve the ponyfill running in an infinite loop. Instead, I've gotten into a state where the MutationObserver enters an infinite loop without ever detecting a mutation in a `<style>` or `<link>` element or firing the ponyfilll. I've updated the stackblitz example to recreate that infinite loop outside of the ponyfill. 

This does seem to be an edge case however, since I can only reproduce it by doing some weird (and probably bad) things with Angular validation. I still think it makes sense to allow the user to dictate what the MutationObserver attaches to, since firing every time the dom changes can be costly. In my application for example, I know that the only css-variables that I want to ponyfill are coming from my own application's code, so it would be completely safe to only watch the `<head>` element.

**Comment by jhildenbiddle** [comment]
@wclaibor --

The infinite loop in your demo is not caused by the ponyfill.

Replace the contents of `ponyfillCssVars.ts` in your demo with the following code which completely removed the ponyfill from the demo:

```js
export function ponyfillCssVars() {
  return () => {
    // The angular application waits for this promise to
    // resolve before starting the application
    return new Promise(resolve => {
      console.log("begin ponyfill");

      // (Ponyfill call removed)
      
      resolve()
    });
  };
}
```

The infinite loop still occurs. I didn't dig into the code too much, but I'm guessing the issue is with the MutationObserver created in `app.component.ts`.
